code/client/Window.cpp

```diff
@@ -4,66 +4,37 @@
-	CallbackInterface* callbacks = static_cast<CallbackInterface*>(glfwGetWindowUserPointer(window));
-	callbacks->keyCallback(key, scancode, action, mods);
-}
-
-
-void Window::mouseButtonMetaCallback(GLFWwindow* window, int button, int action, int mods) {
-	CallbackInterface* callbacks = static_cast<CallbackInterface*>(glfwGetWindowUserPointer(window));
-	callbacks->mouseButtonCallback(button, action, mods);
-}
-
-
-void Window::cursorPosMetaCallback(GLFWwindow* window, double xpos, double ypos) {
-	CallbackInterface* callbacks = static_cast<CallbackInterface*>(glfwGetWindowUserPointer(window));
-	callbacks->cursorPosCallback(xpos, ypos);
-}
-
-
-void Window::scrollMetaCallback(GLFWwindow* window, double xoffset, double yoffset) {
-	CallbackInterface* callbacks = static_cast<CallbackInterface*>(glfwGetWindowUserPointer(window));
-	callbacks->scrollCallback(xoffset, yoffset);
-}
-
-
-void Window::windowSizeMetaCallback(GLFWwindow* window, int width, int height) {
-	CallbackInterface* callbacks = static_cast<CallbackInterface*>(glfwGetWindowUserPointer(window));
-	callbacks->windowSizeCallback(width, height);
+void sdl_error_handler(int sdl_return) {
+	if (!sdl_return) return;
+	Log::error("SDL error: ", SDL_GetError());
+	SDL_Quit();
+	exit(1);
 }
 
 
-// ----------------------
-// non-static definitions
-// ----------------------
-
-Window::Window(
-	std::shared_ptr<CallbackInterface> callbacks, int width, int height,
-	const char* title, GLFWmonitor* monitor, GLFWwindow* share
-)
+Window::Window(int width, int height, const char* title)
 	: window(nullptr)
-	, callbacks(callbacks)
 {
 	// specify OpenGL version
-	glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
-	glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
-	glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
-	glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE); // needed for mac?
-	glfwWindowHint(GLFW_OPENGL_DEBUG_CONTEXT, GL_TRUE);
-
-	// create window
-	window = std::unique_ptr<GLFWwindow, WindowDeleter>(glfwCreateWindow(width, height, title, monitor, share));
-	if (window == nullptr) {
-		Log::error("WINDOW failed to create GLFW window");
-		throw std::runtime_error("Failed to create GLFW window.");
-	}
-	glfwMakeContextCurrent(window.get());
+	// XXX(beau): handle errors?
+	SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 4);
+	SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 6);
+	SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK, SDL_GL_CONTEXT_PROFILE_CORE);
+	SDL_GL_SetAttribute(SDL_GL_DOUBLEBUFFER, 1);
+	SDL_GL_SetAttribute(SDL_GL_CONTEXT_FLAGS, SDL_GL_CONTEXT_DEBUG_FLAG);
+	SDL_GL_SetAttribute(SDL_GL_CONTEXT_FLAGS, SDL_GL_CONTEXT_FORWARD_COMPATIBLE_FLAG);
+
+	window = std::unique_ptr<SDL_Window, WindowDeleter>(SDL_CreateWindow("CPSC 453", SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED, width, height, SDL_WINDOW_OPENGL));
+	if (!window)
+		sdl_error_handler(1);
+
+	// TODO:(beau) clean up
```

Comment:
does this need to be cleaned up still?

Reply:
I spoke too soon. There's some other things we could be doing to clean up sdl as well, but atm having the os clean up when the process ends should work fine. It's another situation where I think where stuff should go will become clear later. I'll put a long note in the window class that enumerates the sdl cleanup things we should do.

e.g. run SDL_Qui() with atexit? with the window destructor? what about cleaning up or opengl context? will we change contexts? clean up in the destructor? also things have to run in a certain order or weird shit happens so we'll probably have to look into that, etc.